### ISSUE TITLE
Generalize styles with optional values.

### ISSUE THREAD

**Issue by oscbyspro** [oscbyspro/DiffableTextViews#136]
The process of enabling optionality is formulaic, assuming the simplest approach: a style that supports optionality provides optional versions of `format(_:with:)`, `interpret(_:with:)` and `resolve(_:with:)`, among which only the last method is conceivably non-trivial. A wrapper then uses these methods. There are of course other ways of doing this, but the best place to implement optionality seems invariably to be next to the 'standard' implementation.

The real question is if this should be opted into with a protocol, or if all styles should have these methods and provide their own implementations as needed. On one hand, the formula is so utterly trivial that another protocol just seems like another headache. But, on the other hand, optional values do not always make sense: `empty` is a more sensible representation, if the value is a `collection`.

**Comment by oscbyspro** [comment]
I suppose my indecision here is because ‘senselessness’ is not a strong argument. `Int???????` is a senseless, but valid type, and nobody has ever needed the compiler to force them into using something more sensible, for example.

**Comment by oscbyspro** [comment]
I'm working on a protocol version of this proposal. It works well, although it is a bit more complicated than baking it all into the main protocol. I think it is correct, however. I'd just like to note that the default `resolve(_:with:)` implementation should be a bit different than proposed in #132. The improved version is as shown below:

```swift
@inlinable func resolve(optional proposal: Proposal,
with cache: inout Cache) throws -> Commit<Value?> {
    //=--------------------------------------=
    // Erase
    //=--------------------------------------=
    if  proposal.replacement.isEmpty,
        proposal.range.lowerBound == proposal.base.startIndex,
        proposal.range.upperBound == proposal.base.endIndex {
        return Commit()
    //=--------------------------------------=
    // Other
    //=--------------------------------------=
    } else { return try Commit(resolve(proposal, with: &cache)) }
}
```

The difference is that this default would work 'decently' for the numbers, as the value would become `nil` when all characters are deleted (cmd + backspace, etc). In the other version 'delete all' would still be interpreted as `$0`.

**Comment by oscbyspro** [comment]
Solved in e6ef5ec4bb23dd203fc1d6bfd93e2bf7b9f7120c with the introduction of the `NullableTextStyle` protocol.
